libvmi/driver/kvm/kvm.c

```diff
@@ -1027,54 +1016,35 @@ kvm_resume_vm(
     kvm_instance_t *kvm = kvm_get_instance(vmi);
 
     // already resumed ?
-    if (!kvm->expected_pause_count)
+    if (!kvm->paused)
         return VMI_SUCCESS;
 
-    // wait to receive pause events
-    while (kvm->expected_pause_count) {
+    // iterate over pause_events_list and unpause every single vcpu by sending a continue reply for each pause event
+    for (unsigned int vcpu = 0; vcpu < vmi->num_vcpus; vcpu++) {
         struct kvmi_dom_event *ev = NULL;
-        unsigned int ev_id = 0;
-
-        // check pause events poped by vmi_events_listen first
-        for (unsigned int vcpu = 0; vcpu < vmi->num_vcpus; vcpu++) {
-            if (kvm->pause_events_list[vcpu]) {
-                dbprint(VMI_DEBUG_KVM, "--Removing PAUSE_VCPU event from the buffer\n");
-                ev = kvm->pause_events_list[vcpu];
-                kvm->pause_events_list[vcpu] = NULL;
-                break;
-            }
-        }
 
-        // if no pause event is waiting in the list, pop next one
-        if (!ev) {
-            if (VMI_FAILURE == kvm_get_next_event(kvm, &ev, 1000)) {
-                errprint("Failed to get next KVMi event\n");
-            }
-            if (!ev) {
-                // no new events
-                // report error
+        // wait until the pause event for the current vcpu is received
+        while (!kvm->pause_events_list[vcpu]) {
```

Comment:
I wonder: Maybe it makes sense to wait for all pause events before starting to unpause vcpus? Or is it irrelevant that there may be a couple of milliseconds between two `reply_continue()` calls? :thinking:

Reply:
I don't think it's necessary to wait for all VCPU pause events to be received here
if you have 64 VCPUs, you want to resume them ASAP.